packages/starlight/locals.ts

```diff
@@ -1,8 +1,38 @@
+import type { APIContext } from 'astro';
+import { AstroError } from 'astro/errors';
 import { defineMiddleware } from 'astro:middleware';
+import type { StarlightRouteData } from './route-data';
 import { useTranslations } from './utils/translations';
 
-export const onRequest = defineMiddleware((context, next) => {
+export const onRequest = defineMiddleware(async (context, next) => {
 	context.locals.t = useTranslations(context.currentLocale);
-
+	initializeStarlightRoute(context);
 	return next();
 });
+
+/**
+ * Sets up a `starlightRoute` property on locals. Initially, this will throw an error if accessed.
+ * When rendering, Starlight’s routes set `starlightRoute` with the resolved route data object for
+ * the current page.
+ *
+ * This ensures Starlight components can easily access `starlightRoute` without needing type guards,
+ * we can throw a helpful message if `starlightRoute` is accessed on non-Starlight pages, and we
+ * avoid generating route data in this middleware which also runs for non-Starlight route.
+ */
+export function initializeStarlightRoute(context: APIContext) {
+	const state: { routeData: StarlightRouteData | undefined } = { routeData: undefined };
+	Object.defineProperty(context.locals, 'starlightRoute', {
+		get() {
+			if (!state.routeData) {
```

Comment:
Is a `try/catch` like you had in [`13d2038` (#2390)](https://github.com/withastro/starlight/pull/2390/commits/13d203846592f23b23120d6236f19fa9b389bd83) the recommended approach to check if it's defined? We don't provide any helper or boolean to avoid a `try/catch`?

I'm fine with that, just wondering.

I assume we should maybe document this behavior too in `docs/src/content/docs/reference/route-data.mdx`?

Reply:
Hmm, good question! What are the scenarios/use cases? Components that should behave in a certain way inside Starlight routes but also work in other places?

Reply:
I guess this is now be way less common/important in this new implementation vs the previous one, considering now a route middleware will only run for Starlight routes which was not the case before.

Considering route data is now no longer in `Astro.props` but in `Astro.locals`, it allows to basically access the route data in lot of new places, so yeah, could be in a user-component and based on the presence of route data, you could decide to do something different. I could even imagine scenarios where you could check for it in another regular middleware.

I'm fine with the current approach if you need to check for it, just think it should be noted in the documentation that this can potentially throw an error.

Reply:
I think I’m OK with requiring a `try/catch` for now? In the future we could always consider a `hasStarlightRoute` boolean or something? Although I don’t love if that would _maybe_ be confusing to other middleware? It would only indicate `starlightRoute` has been attached, _not_ whether it _might_ be attached later I guess.